### PR TITLE
Pin Java version + add support for Java 9, 10, 11

### PR DESCRIPTION
**Pull request by hadim**
@hanslovsky: Here is a first try to pin the Java version.

### CODE REVIEW

**Comment by hadim** [comment]
@jakirkham & @sodre: does this approach make sense to you?

**Comment by sodre** [comment]
@hadim, if you are trying to build `pyjnius` against different versions of `java`, e.g. 8, 9, 10, 11, etc.. here is how I did it within the context of `postgis` and `postgres`.

1. I kept the `postgis` [version](https://github.com/conda-forge/postgis-feedstock/blob/2491ec60c243e1478e4e83f084466bdb1d39b116/recipe/meta.yaml#L2) and [build](https://github.com/conda-forge/postgis-feedstock/blob/2491ec60c243e1478e4e83f084466bdb1d39b116/recipe/meta.yaml#L16-L19) strings intact as you would normally do with any package.

2. I added `postgres` as a [host](https://github.com/conda-forge/postgis-feedstock/blob/2491ec60c243e1478e4e83f084466bdb1d39b116/recipe/meta.yaml#L32) and [run time](https://github.com/conda-forge/postgis-feedstock/blob/2491ec60c243e1478e4e83f084466bdb1d39b116/recipe/meta.yaml#L40) dependencies in the meta.yaml file without any version information.

3. I created the [conda_build_config.yaml](https://github.com/conda-forge/postgis-feedstock/blob/master/recipe/conda_build_config.yaml) file pinning the `postgres` versions.

4. Finally, I reran the `conda smithy rerender` command to get the additional ci_support files for [CircleCI and Travis](https://github.com/conda-forge/postgis-feedstock/tree/master/.ci_support).

**Comment by sodre** [comment]
I just saw your conda_build_config.yaml and meta.yaml files. I think you just need to add the max_pin to conda_build_config.yaml, remove the build.string field, and rerun codna smithy rerender to get this to work.

File: recipe/conda_build_config.yaml
Comment (sodre) [inline recipe/conda_build_config.yaml:2]: If pyjnius works with Java 9 and you add that here, then you will get twice as many builds and would help verify that what I think you want to do is actually working.

Comment (hadim) [reply recipe/conda_build_config.yaml:2]: I know but at the moment `pyjnius` is not compatible with Java >=9. We need https://github.com/kivy/pyjnius/issues/304 to be fixed first.

**Comment by hadim** [comment]
@hanslovsky : ready for review/merge.

**Comment by hadim** [comment]
Good news: the next version of pyjnius will be compatible with Java >=9: https://github.com/kivy/pyjnius/pull/363

**Comment by hadim** [comment]
This is a work in progress: DO NOT MERGE.

**Comment by hadim** [comment]
Ok, all the builds should turn green at some point.

Before merging this we need pyjnius to release a new version since at the moment the recipe uses pyjnius `master` branch: https://github.com/kivy/pyjnius/pull/363#issuecomment-441376564

@mingwandroid & @jakirkham & @sodre : I would like to check with you I am not doing a mistake with compilers here.

By default on linux and osx, builds happen on two compilers but for a reason I don't understand `openjdk` is only available for one of them called `toolchain_c` so I had to fix the compilers used in `conda_build_config.yaml` for both linux and osx:

```yaml
openjdk:
  - 8
  - 9
  - 10
  - 11

pin_run_as_build:
  openjdk:
    max_pin: x.x

c_compiler:
  - toolchain_c  # [not win]
  - vs2015  # [win]
```

Can you confirm whether is it ok to do it that way? Thank you!

**Comment by hanslovsky** [comment]
> ready for review/merge.

I do not think that I can review a lot here as my conda knowledge is very shallow. Green checkmarks for the build and thumbs-up from one or more experts will suffice for me to merge this.

> This is a work in progress: DO NOT MERGE.

Gotcha

**Comment by mariusvniekerk** [comment]
So not adding the compilers from pinnings may cause issues as we are moving away from things compiled with gcc 4.8 in favor of the anaconda compilers.

What exactly is missing?  are there sine versions of openjdk missing for the new compilers?

**Comment by sodre** [comment]
@mariusvniekerk,
the main problem is that openjdk is a binary repackage and thus did not use 'toolchain' or the anaconda compilers.

To use the Anaconda compilers, @hadim would have to build the openjdk from scratch. I am not sure how easy/feasible it is to do that.

**Comment by hadim** [comment]
I actually don't know what is missing on the OpenJDK side since it does not depend on a compiler. A compiler is only used for testing the build: https://github.com/conda-forge/openjdk-feedstock/blob/master/recipe/meta.yaml#L33

**Comment by hadim** [comment]
But since it does not depend on a compiler, why it is not available on some of them?

**Comment by sodre** [comment]
I think it is because it has not gone through the GCC7 Migration yet, e.g. the build number is <1000

**Comment by hadim** [comment]
Ok, so it's an easy fix I guess.

Can I migrate myself? Or do I have to wait for the bot?

**Comment by hadim** [comment]
Should I set `noarch: python` to the openjdk recipe?

**Comment by hadim** [comment]
I checked https://conda-forge.org/status/ and couldn't find `openjdk` in it. I guess it's because no compiler is declared in its recipe.

As already said the compiler is only declared in [the test section](https://github.com/conda-forge/openjdk-feedstock/blob/e6e3105ae4420b841465f52f728). Is that why a compiler is declared in the [yaml configuration](https://github.com/conda-forge/openjdk-feedstock/blob/e6e3105ae4420b841465f52f728c91246d305cd7/.ci_support/linux_.yaml#L1)?

A solution would be to remove tests that use C compiler, that would make the package compiler-independent. On the other side, I think it's a good practice to check that JNI compilation works.

**Comment by hadim** [comment]
This PR should fix the issue https://github.com/conda-forge/openjdk-feedstock/pull/42

Once it's merged I will remove the hard-coded C compilers from this PR.

**Comment by hadim** [comment]
To do before merging here:

- [x]  Java 8: https://github.com/conda-forge/openjdk-feedstock/pull/43
- [x]  Java 9: https://github.com/conda-forge/openjdk-feedstock/pull/44
- [x]  Java 10: https://github.com/conda-forge/openjdk-feedstock/pull/45
- [x]  Java 11: https://github.com/conda-forge/openjdk-feedstock/pull/46
- [x] Add more Java version to this recipe.
- [ ] Wait for pyjnius to release a new version.

**Comment by hadim** [comment]
It's almost good. Only OSX builds for `toolchain_c` are failing when solving env:

```
conda.exceptions.ResolvePackageNotFound: 
  - openjdk=10
```

It's weird because conda-forge has those builds. Example for Java 11: https://anaconda.org/conda-forge/openjdk/files?version=11.0.1. There is OSX builds for both `toolchain_c` and `clang` with `main` and `gcc7` labels. I also checked inside those packages and `info/hash_input.json` contains (for the `main` labeled package):

```json
{
  "c_compiler": "toolchain_c"
}
```

Does Anaconda repos have an index that could have been confused by the java branches I have created on the Java feedstock repo? See https://github.com/conda-forge/openjdk-feedstock/branches

**Comment by hadim** [comment]
I guess conda servers have been fixed because the missing OSX OpenJDK packages are back.

It's all green now.

Now waiting for a pyjnius release.

File: recipe/conda_build_config.yaml
Comment (isuruf) [inline recipe/conda_build_config.yaml:4]: Why support jdk 9 and 10? They have reached end of support.

Comment (hadim) [reply recipe/conda_build_config.yaml:4]: Projects such as Imagej does not support Java 10 and 11 yet while it supports Java 9 which has modern feature Java 8 does not have.

Conda OpenJDK builds are available so I don't see anything bad to support it. Moreover, by default with `conda install openjdk`, people will get the latest Java 11 version.

**Comment by hadim** [comment]
@sodre could you review, please?

File: recipe/meta.yaml
Comment (sodre) [inline recipe/meta.yaml]: I am out of the loop as far as how the auto-tick bot works, but I think it is probably better to revert this back to 1001,  and remove line 3.

File: recipe/meta.yaml
Comment (sodre) [inline recipe/meta.yaml]: Is the goal to still wait for a proper release to PyPi? If so, this needs to be updated.

**Review by sodre — CHANGES_REQUESTED** [review]
@hadim this looks fine to me minus the build_number jinja2 expression and the fact that we are still pulling from github. 

Keep in mind that I am not a maintainer of this package so I can't really merge this into the master branch.

File: recipe/meta.yaml
Comment (hanslovsky) [reply recipe/meta.yaml]: As far as I understand we cannot use the PyPI release because it does not ship the `pyjnius.jar` file that we need to implement Java interfaces in Python. We are still waiting for a proper tag on GithHub so we can have a proper version string and download from a GH release.

File: recipe/meta.yaml
Comment (hadim) [reply recipe/meta.yaml]: Will do.

File: recipe/meta.yaml
Comment (hadim) [reply recipe/meta.yaml]: That is exact, we will download from GH and still use sha256. We need to wait for a git tag on the pyjnius side.

**Comment by hadim** [comment]
@sodre Is that ok to release a `dev` version in `master` here? I feel it could take a while before we see a git tag on the `pyjnius` side.

Something like `1.1.3-dev0-GIT_HASH`?

**Comment by sodre** [comment]
> @sodre Is that ok to release a `dev` version in `master` here? I feel it could take a while before we see a git tag on the `pyjnius` side.
> 
> Something like `1.1.3-dev0-GIT_HASH`?

That is probably a question better asked to @conda-forge/core ...

**Comment by hadim** [comment]
I got a green light on gitter from @CJ-Wright to merge here.

@hanslovsky only can merge this PR and I know he is AFK for two weeks.

@CJ-Wright: could you merge this PR (if you can...)?

**Comment by CJ-Wright** [comment]
@hadim any interest in joining the maintainers of this package?

**Comment by hadim** [comment]
I will be once this PR is merged :-)

**Comment by CJ-Wright** [comment]
sorry I missed that bit

**Pull request merged**
